### ISSUE TITLE
(One shot) manual data display validation

### ISSUE THREAD

**Issue by niconoe** [inbo/crow#45]
It would be nice to once load a tiny subset of the data (like 15 minutes at 6 different height) so we can manually check the heatmap works correctly for things such as:

- data correctness (so far we only now it "looks like BioRad output")
- one-off errors
- padding/margins between rectangles (so they don't overlap, don't walk on the axis shoes, have a nice padding a different timescale, ...)

---

- [x] Provide good sample days @mrtnrey : 2020-10-12/14 (by @robinbrabant)
- [x] Create Rmd to read into bioRad and create charts @peterdesmet , see https://github.com/inbo/crow/blob/master/tests/crow_biorad_comparison.Rmd
- [x] Compare charts with crow @mrtnrey @peterdesmet

**Comment by peterdesmet** [comment]
Agreed. And check if the calculated MTR values are actually correct (by comparing with bioRad output)

**Comment by mrtnrey** [comment]
MTRs are very (too?) high in my opinion. Should be checked indeed.

**Comment by peterdesmet** [comment]
To be tested (ideally by @mrtnrey) after #38 #76 are fixed.

**Comment by peterdesmet** [comment]
Here is a comparison (top chart: CROW, bottom chart: bioRad):

![crow_plot](https://user-images.githubusercontent.com/600993/107504657-f20d6880-6b9b-11eb-8ac2-8b566ba3ad01.png)
![biorad_plot](https://user-images.githubusercontent.com/600993/107504664-f3d72c00-6b9b-11eb-9763-6766d44ccb7e.png)

**Comment by peterdesmet** [comment]
The alignment of height and time looks perfect between the two charts, but the colourscale is off: 50 in bioRad is still in the red zone, while in CROW it is purple.

**Comment by peterdesmet** [comment]
Here is the comparison for the VPI (MTR) data (top chart: CROW, bottom chart: bioRad):

![crow_vpi](https://user-images.githubusercontent.com/600993/107509758-48ca7080-6ba3-11eb-8ce1-6538d8b06a8f.png)
![biorad_vpi](https://user-images.githubusercontent.com/600993/107509773-4cf68e00-6ba3-11eb-9eea-525cf8ec531d.png)

**Comment by peterdesmet** [comment]
Again: the x and y alignment of the data looks perfect between the two charts. IMO we can conclude that the charts are ok, pending the issue of the bioRad colour scale.

**Comment by mrtnrey** [comment]
Whew! Nice to see the agreement.

Concerning the colour scale difference: the bioRad scale for CROW seems stretched till 250 (max value) while max value in the bioRad scale in (R-bioRad) seems to be 1000. Can this explain the difference, or is this naive?

**Comment by niconoe** [comment]
@mrtnrey : I made a quick and dirty test, and I think this is indeed the reason behind the difference: CROW takes the max density as the top of the scale, while BioRAD use a fixed value of 1000.

I guess both have their advantages: BioRAD allows direct comparison between very different days, while the CROW approach probably gives more nuances on slow days.

But if the main goal is to be consistent with BioRAD, I can adapt this for this color scale (but I suggest keeping the other approach for the Viridis scale so we have complementary solutions there: 1 scale for nuances, 1 scale for comparing values).

any feedback?

**Comment by peterdesmet** [comment]
@niconoe that seems perfect to me. I think we should match the bioRad one as close as possible.

**Comment by robinbrabant** [comment]
I agree. Very nice to see that the plots are matching that well and that MTR values are identical. The BioRAD MTR plot has some more detail, is that because MTR is calculated on shorter time frames?

**Comment by peterdesmet** [comment]
@robinbrabant that is correct, the data is at a 5min resolution (which is used as is by bioRad), while CROW samples at [10 min](https://github.com/inbo/crow/blob/78783334685ecc9c2998081f1dd47d7374147867/src/config.ts#L72) to make the charts about twice as fast.

**Comment by mrtnrey** [comment]
Allow me to propose a third solution: a "soft" max :smile:
You could set the max to 250, unless there are values bigger than that.

The bioRad scale is not carved in stone, so we might suggest to change it in the bioRad repo instead.

**Comment by mrtnrey** [comment]
But I'm also fine if you stetch it (fixed) to 1000 to have perfect agreement. No problem!

**Comment by peterdesmet** [comment]
@mrtnrey yeah, given that we have ample flexibility with our default Viridis one and that we call it the "bioRad colour scale", I would match it closely. Whether the 1000 max is a good bioRad decision or not, that is how people have plotted charts for a while now. 😄

**Comment by mrtnrey** [comment]
Good! go ahead.

**Comment by niconoe** [comment]
I made a few changes to the bioRAD colour scale that gets us quite close to real thing:

![Screenshot 2021-02-10 at 20 56 25](https://user-images.githubusercontent.com/386387/107564363-9914f300-6be2-11eb-8a3a-320a11e4bc24.png)

[see it yourself](https://inbo.github.io/crow/#/?radar=bewid&date=2020-10-12&interval=72&timedisplay=radarLocal&vpColorScheme=biorad&vpiMode=mtr)

- scale max is always 1000, as discussed above
- the ticks under the X axis shows the same value than bioRad: 1, 2, 5, 10, 25, ...

There's still a tiny colour difference, but I think the reason is that we don't use a real log scale, but a [symlog](https://observablehq.com/@d3/continuous-scales#scale_symlog) one. One easy empirical improvement could be to tweak the `constant` parameter of the scale, to adjust the slope around 0 (see link above).

A cleaner solution would be to check exactly what bioRad does and reimplement. I'm quite bad at this so a second pair of eyes would be great, but I have the impression that they [apply a "real" log scale](https://github.com/adokter/bioRad/blob/33ec6ed50f64029b7cc097b819aa1c9afa0587d4/R/plot.vpts.R#L198-L202) but also that before that, [they push very small values to 0.5](https://github.com/adokter/bioRad/blob/33ec6ed50f64029b7cc097b819aa1c9afa0587d4/R/plot.vpts.R#L130) to avoid having zeroes in data (my previous attempts to use a log scale in CROW failed, probably because I attempted log(0)).

Does any of this makes sense? Should I attempt to mimic this (supposed) bioRad behaviour?

**Comment by mrtnrey** [comment]
Thanks a lot @niconoe !

You're correct that some tiny differences still remain, but - on top of what you discuss above - don't forget that
- the different time resolution (5min↔ 10min) can play a role
- the exact colour handling (colour profile etc) between R as a desktop application and CROW as a browser application

In my opinion we can assign a "low priority" to this ticket at this stage, and first solve the remaining issues prior to launch.

**Comment by peterdesmet** [comment]
Thank @niconoe. I have screenshotted the new fixed 1000 scale chart you made and tested the 5 min resolution. The result is in my comment https://github.com/inbo/crow/issues/45#issuecomment-776645936 (middle chart). The differences are minor, but i'm fine leaving this issue open if you want to take another stab at it.

**Comment by peterdesmet** [comment]
Closing this issue (which is about the validation and resolved) in favour of #45 (which is about the bioRad colour scale)
